Use 2 command bufs on Prefetch D/HD

nhuang-tt opened the pull request:
### Ticket
#20243

### Problem description
Saves the register write by using the stateful API.

### What's changed
- Prefetch D/HD is only writing to `downstream_noc_xy` and `dispatch_s_noc_xy`. We can setup 2 command buf's NoC destination to point to those downstream addresses once during init.
- Write cmd buf allocation:
    - BRISC_WR_CMD_BUF: writes to downstream_noc_xy
    - BRISC_WR_REG_CMD_BUF: small writes to dispatch_s_noc_xy. not much traffic on this path.

-  Using the normal NoC APIs for writes and/or inline_dw_writes are not allowed on this kernel because it's using the stateful APIs now.
- Device profiler to save and restore some NoC destination/cmd registers when it's running on a dispatch kernel

Device 0 test_pgm_dispatch (~7% better)
```
Consider adjusting baselines. Test BM_pgm_dispatch/brisc_only_trace/256/manual_time got value 2.58us but expected 2.78us (7.21% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/brisc_only_trace/512/manual_time got value 2.58us but expected 2.79us (7.53% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/brisc_only_trace/1024/manual_time got value 2.60us but expected 2.82us (7.70% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/brisc_only_trace/2048/manual_time got value 2.73us but expected 2.98us (8.15% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/ncrisc_only_trace/256/manual_time got value 2.58us but expected 2.78us (7.24% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/ncrisc_only_trace/512/manual_time got value 2.58us but expected 2.79us (7.53% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/ncrisc_only_trace/1024/manual_time got value 2.60us but expected 2.82us (7.65% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/ncrisc_only_trace/2048/manual_time got value 2.74us but expected 2.90us (5.40% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/brisc_trisc_only_trace/256/manual_time got value 2.97us but expected 3.17us (6.45% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_trace/1024/manual_time got value 3.32us but expected 3.51us (5.59% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_1cb/512/manual_time got value 3.42us but expected 3.64us (6.09% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_1_core_128_crta/1024/manual_time got value 3.72us but expected 3.93us (5.38% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_128_crta/1024/manual_time got value 3.72us but expected 4.00us (6.88% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/sems_1_core_1_processor_trace/256/manual_time got value 2.91us but expected 3.10us (6.22% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/sems_1_core_1_processor_trace/512/manual_time got value 2.91us but expected 3.10us (6.21% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/sems_1_core_1_processor_trace/1024/manual_time got value 2.91us but expected 3.13us (7.07% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/sems_all_cores_1_processor_trace/256/manual_time got value 2.91us but expected 3.10us (6.21% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/sems_all_cores_1_processor_trace/512/manual_time got value 2.91us but expected 3.10us (6.23% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/sems_all_cores_1_processor_trace/1024/manual_time got value 2.91us but expected 3.13us (7.09% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/sems_all_cores_1_processor_trace/2048/manual_time got value 3.04us but expected 3.27us (7.03% better) .
Consider adjusting baselines. Test BM_pgm_dispatch_vary_slow_cycles/256_bytes_brisc_only_all_processors_trace/0/manual_time got value 2.58us but expected 2.78us (7.26% better) .
Consider adjusting baselines. Test BM_pgm_dispatch_vary_slow_cycles/256_bytes_brisc_only_all_processors_trace/1000/manual_time got value 2.58us but expected 2.78us (7.29% better) .
Test successful
```

Device 1 test_pgm_dispatch (5-7% better)
```
Consider adjusting baselines. Test BM_pgm_dispatch/brisc_only_trace/1024/manual_time got value 2.59us but expected 2.74us (5.36% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/brisc_only_trace/2048/manual_time got value 2.74us but expected 2.92us (5.99% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/ncrisc_only_trace/1024/manual_time got value 2.60us but expected 2.78us (6.57% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/ncrisc_only_trace/2048/manual_time got value 2.73us but expected 2.93us (6.70% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/trisc_only_trace/4096/manual_time got value 3.68us but expected 3.89us (5.26% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_trace/256/manual_time got value 3.07us but expected 3.24us (5.36% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_1cb/1024/manual_time got value 3.60us but expected 3.87us (7.16% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_32cb/1024/manual_time got value 3.61us but expected 3.91us (7.59% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_1cb_1sem/512/manual_time got value 3.40us but expected 3.59us (5.14% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_1cb_1sem/1024/manual_time got value 3.60us but expected 3.87us (7.04% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_1_core_1_rta/1024/manual_time got value 3.62us but expected 3.87us (6.56% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_1_core_1_crta/1024/manual_time got value 3.60us but expected 3.87us (7.06% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_1_core_128_crta/1024/manual_time got value 3.71us but expected 3.94us (5.70% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_1_crta/1024/manual_time got value 3.60us but expected 3.88us (7.18% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_32_crta/1024/manual_time got value 3.60us but expected 3.89us (7.26% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_128_crta/256/manual_time got value 3.46us but expected 3.65us (5.09% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/all_processors_all_cores_128_crta/1024/manual_time got value 3.71us but expected 3.94us (5.69% better) .
Consider adjusting baselines. Test BM_pgm_dispatch/sems_1_core_1_processor_trace/1024/manual_time got value 2.92us but expected 3.08us (5.09% better) .
Consider adjusting baselines. Test BM_pgm_dispatch_vary_slow_cycles/256_bytes_brisc_only_all_processors_trace/1000/manual_time got value 2.60us but expected 2.75us (5.41% better) .
```

### Checklist
APC
https://github.com/tenstorrent/tt-metal/actions/runs/14672544503
https://github.com/tenstorrent/tt-metal/actions/runs/14672543313
T3K
https://github.com/tenstorrent/tt-metal/actions/runs/14674087209
BH
https://github.com/tenstorrent/tt-metal/actions/runs/14650127736
test_pgm_dispatch
https://github.com/tenstorrent/tt-metal/actions/runs/14650834834

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -202,15 +202,16 @@ FORCE_INLINE void write_downstream(
     if (length > remaining) {
         if (remaining > 0) {
-            noc_async_write(
+            cq_noc_async_write_with_state_any_len<true, true>(
```
looks like this fn is "inline" not "force_inline", wonder if that matters

pgkeller replied:
this can be optimized but not updating the counters in the loop but instead at the end

nhuang-tt replied:
> looks like this fn is "inline" not "force_inline", wonder if that matters

I verified that it's getting inlined

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_prefetch.cpp:
```diff
@@ -398,6 +399,14 @@ static uint32_t process_relay_inline_cmd(uint32_t cmd_ptr, uint32_t& local_downs
+    // By default we initialized downstream to downstream_noc_xy
+    // If RelayInlineState is a DispatchSRelayInlineState, we need to initialize downstream to dispatch_s_noc_xy
+    // and then clean up after
+    if constexpr (std::is_same_v<RelayInlineState, DispatchSRelayInlineState>) {
```
definitely think we can gain by a full spec of use of noc+cmd bufs so we can dedicate a cmd buf for writing to dispatch_[d/s].  we can start by enumerating our use cases (in slack), maybe we should meet w/ John/Mihir later this week (or resolve over slack)

pgkeller reviewed (commented):
conceptually this looks fine though perhaps we want to wait to define which cmd bufs are used for what so we don't have to override/restore state at the time of use, this could be a maintenance issue

nhuang-tt committed:
Add noc_index parameters to cq_common

nhuang-tt committed:
Utilize 2 command bufs in prefetch_d/hd

7% decrease in dispatch time for many kernel groups by
using 1 cmd buf for writing to dispatch_d and another for
writing to dispatch_s

nhuang-tt committed:
Update pgm_dispatch_golden.json

nhuang-tt committed:
Add cmd_buf parameter to DEBUG_SANITIZE_NOC_WRITE_TRANSACTION_FROM_STATE

mpiseTT commented on tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/pgm_dispatch_golden.json:
Some help to understand these results. I'm seeing generally `iterations` increasing a bit and `iteration time` decreasing a bit. There are also several instances where the latter has increased. The `real time` seems to be almost the same as `iteration time` while the `cpu time` is lower in almost every if not all cases. I guess the `real time` or `iteration time` are the metrics that are experienced by the user.

nhuang-tt replied:
Google benchmark will run the test case for number of iterations until the results become stable.
IterationTime is what we report from our test

jbaumanTT reviewed (approved):
Seems good. I'm working on a change that will allow us to enforce that the standard dataflow apis aren't accidentally used, which should make this more maintainable.

nhuang-tt committed:
Save/Restore NoC destination state in profiler

nhuang-tt commented:
@mo-tenstorrent There's a conflict in `kernel_profiler.hpp`. Prefetcher has some NoC registers setup only once in kernel main. The profiler pushing is changing these registers. I added some code in the profiler to save/register the affected registers for dispatch kernels.
